activerecord/lib/active_record/associations/has_many_through_association.rb

```diff
@@ -53,16 +53,24 @@ def concat_records(records)
         #
         # However, after insert_record has been called, the cache is cleared in
         # order to allow multiple instances of the same record in an association.
-        def build_through_record(record)
-          @through_records[record] ||= begin
-            ensure_mutable
+        def through_records(record)
+          @through_records[record] ||= build_through_record(record)
+        end
 
-            attributes = through_scope_attributes
-            attributes[source_reflection.name] = record
-            attributes[source_reflection.foreign_type] = options[:source_type] if options[:source_type]
+        def build_through_record(record)
+          ensure_mutable
 
-            through_association.build(attributes)
+          if (inverse = source_reflection.inverse_of) && inverse.has_one? &&
+             (has_one_target = record.association(inverse.name).scope.where(through_scope_attributes).first)
```

Comment:
This is a pretty complex query and it's a bit unclear to me what it is loading.
Is there maybe a easier way to load this?
Maybe something like:
https://github.com/rails/rails/blob/f5b74ff2ac298c11f06cd86120fc9af78074a68d/activerecord/lib/active_record/associations/has_one_through_association.rb#L19
I think we should also move it outside the `if` statement.

Reply:
Readability fixed be the separate method, I guess.

I think that `through_record = through_proxy.load_target` wouldn't work in all the cases, we have (from test schema):
`Post.has_many: :single_people` -> `Reader has_one :single_person` -> `Person.has_one :reader` & `Person.has_many :posts, through: :readers`
and here we're doing (pseudocode)
```ruby
inverse = SourceReflection.new(:single_person).inverse_of # => :reader from `Person.has_one :reader`
has_one_target_association = Person.association(:reader) # Reader.where(person_id: reader.id)
# Consider `Post.has_many: :single_people, -> { where(foo: bar) }` case
# Loads `Reader.where(person_id: person.id).first`
has_one_target = has_one_target_association.load_target
# Loads `Reader.where(person_id: person.id).where(foo: :bar).first`
has_one_target = has_one_target_association.scope.where(through_scope_attributes).first 
```
I can try to craft a testcase for that, although probably another one `Post -> Reader -> Person` association chain is needed.

Reply:
naming is hard, hopefully method/variable is named more-or-less understandable